### PR TITLE
BTreeIndex Charge Keys

### PR DESCRIPTION
**Pull request by kungasc**

### CODE REVIEW

File: ydb/core/tablet_flat/flat_part_charge_btree_index.h [hunk 1]
```diff
@@ -11,56 +11,107 @@ class TChargeBTreeIndex : public ICharge {
     using TBtreeIndexMeta = NPage::TBtreeIndexMeta;
     using TRecIdx = NPage::TRecIdx;
     using TGroupId = NPage::TGroupId;
+    using TChild = TBtreeIndexNode::TChild;
+
+    // TODO: store PageId only instead of TChild?
+    struct TNodeState : TBtreeIndexNode, TChild {
```
Comment (kunga) [inline ydb/core/tablet_flat/flat_part_charge_btree_index.h:17]: Так нормально, или лучше TBtreeIndexNode и TChild полями сделать?

File: ydb/core/tablet_flat/flat_part_charge_range.cpp [hunk 2]
```diff
@@ -29,30 +23,32 @@ bool ChargeRange(IPages *env, const TCells key1, const TCells key2,
...
-        if (cmp >= 0 /* slice->LastKey >= key2 */) {
+        if (cmp >= 0) {
+            // key2 <= LastKey
             if (r.Overshot && ++pos != run.end()) {
-                // Unfortunately key > key2 might be at the start of the next slice
+                // Unfortunately first key > key2 might be at the start of the next slice
                 TRowId firstRow = pos->Slice.BeginRowId();
                 // Precharge the first row on the next slice
-                CreateCharge(env, *pos->Part, tags, includeHistory)->Do(firstRow, firstRow, keyDefaults, items, bytes);
+                ready &= CreateCharge(env, *pos->Part, tags, includeHistory)->Do(firstRow, firstRow, keyDefaults, items, bytes);
```
Comment (kunga) [inline ydb/core/tablet_flat/flat_part_charge_range.cpp:44]: Не понимаю как оно раньше работало..

File: ydb/core/tablet_flat/flat_part_charge_btree_index.h [hunk 3]
```diff
@@ -11,56 +11,110 @@ class TChargeBTreeIndex : public ICharge {
...
-                    if (i + 1 == level.size() && row2 < meta.RowCount) {
-                        to = level[i].Seek(row2);
-                    }
-                    for (TRecIdx j : xrange(from, to + 1)) {
-                        ready &= TryLoadNode(level[i].GetShortChild(j).PageId, nextLevel);
+        TVector<TNodeState> level(Reserve(3)), nextLevel(Reserve(3));
+        TPageId key1PageId = key1 ? meta.PageId : Max<TPageId>();
+        TPageId key2PageId = key2 ? meta.PageId : Max<TPageId>();
+
+        const auto iterateLevel = [&](std::function<bool(TNodeState& current, TRecIdx pos)> tryLoadNext) {
```
Comment (snaury) [inline ydb/core/tablet_flat/flat_part_charge_btree_index.h:70]: I would recommend making the argument `const auto& tryLoadNext` here. You are passing lambdas below, and packaging them into std::function may be expensive or even allocate memory depending on state size.

File: ydb/core/tablet_flat/flat_part_charge_btree_index.h [hunk 4]
```diff
@@ -70,66 +124,89 @@ class TChargeBTreeIndex : public ICharge {
...
-                    if (i + 1 == level.size() && row2 < meta.RowCount) {
-                        to = level[i].Seek(row2);
-                    }
-                    for (TRecIdx j = from + 1; j > to; j--) {
-                        ready &= TryLoadNode(level[i].GetShortChild(j - 1).PageId, nextLevel);
+        TVector<TNodeState> level(Reserve(3)), nextLevel(Reserve(3));
+        TPageId key1PageId = key1 ? meta.PageId : Max<TPageId>();
+        TPageId key2PageId = key2 ? meta.PageId : Max<TPageId>();
+
+        const auto iterateLevel = [&](std::function<bool(TNodeState& current, TRecIdx pos)> tryLoadNext) {
```
Comment (snaury) [inline ydb/core/tablet_flat/flat_part_charge_btree_index.h:162]: Use `const auto& tryLoadNext` here too.

**Review by snaury — CHANGES_REQUESTED** [review]

**Review by snaury — APPROVED** [review]

**Pull request merged**